src/librustc_mir/borrow_check/mod.rs

```diff
@@ -677,9 +732,24 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
             context,
             (sd, place_span.0),
             flow_state,
-            |this, _index, borrow| match (rw, borrow.kind) {
-                (Read(_), BorrowKind::Shared) => Control::Continue,
-                (Read(kind), BorrowKind::Unique) | (Read(kind), BorrowKind::Mut) => {
+            |this, index, borrow| match (rw, borrow.kind) {
+                // Obviously an activation is compatible with its own reservation;
+                // so don't check if they interfere.
+                (Activation(_, activating), _) if index.is_reservation() &&
```

Comment:
Make sure you don't break this pattern:
```Rust
let mut a = 0;
let mut b = 0;
let p = if maybe {
    &mut a
} else {
    &mut b
};
use(p);
```

Reply:
(added a test for this; see two-phase-control-flow-split-before-activation.rs )